client/mm/wrapped_cex.go

```diff
@@ -0,0 +1,294 @@
+		}
+		w.CEX.ConfirmDeposit(ctx, txID, conf)
+	}()
+
+	return nil
+}
+
+func (w *wrappedCEX) Withdraw(ctx context.Context, assetID uint32, amount uint64, onConfirm func()) error {
+	symbol := dex.BipIDSymbol(assetID)
+
+	balance := w.mm.botCEXBalance(w.botID, assetID)
+	if balance < amount {
+		return fmt.Errorf("bot has insufficient balance to withdraw %s. required: %v, have: %v", symbol, amount, balance)
+	}
+
+	addr, err := w.mm.core.NewDepositAddress(assetID)
+	if err != nil {
+		return err
+	}
+
+	conf := func(withdrawnAmt uint64, txID string) {
+		go func() {
+			checkTransaction := func() bool {
+				_, err := w.mm.core.TransactionConfirmations(assetID, txID)
+				if err == nil {
+					// Assign to balance to the bot as long as the wallet
+					// knows about the transaction.
+					w.mm.modifyBotBalance(w.botID, []*balanceMod{{balanceModIncrease, assetID, balTypeAvailable, withdrawnAmt}})
+					onConfirm()
+					return true
+				}
+				if !errors.Is(err, asset.CoinNotFoundError) {
+					w.log.Errorf("error checking transaction confirmations: %v", err)
+				}
+				return false
+			}
+
+			if checkTransaction() {
+				return
+			}
+
+			ticker := time.NewTicker(time.Second * 20)
+			giveUp := time.NewTimer(time.Minute * 10)
+			defer ticker.Stop()
+			defer giveUp.Stop()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					if checkTransaction() {
+						return
+					}
+				case <-giveUp.C:
+					w.log.Errorf("timed out waiting for withdrawal confirmation")
+					return
+				}
+			}
+		}()
+	}
+
+	err = w.CEX.Withdraw(ctx, assetID, amount, addr, conf)
+	if err != nil {
+		return err
+	}
+
+	w.mm.modifyBotCEXBalance(w.botID, assetID, amount, balanceModDecrease)
+	return nil
+}
+
+func (w *wrappedCEX) handleTradeUpdate(update *libxc.TradeUpdate) {
+	w.tradesMtx.Lock()
+	defer w.tradesMtx.Unlock()
+
+	trade, found := w.trades[update.TradeID]
+	if !found {
+		w.log.Errorf("wrappedCEX: trade ID %s not found", update.TradeID)
+		return
+	}
+
+	if trade.sell && update.QuoteFilled > trade.quoteFilled {
+		quoteFilledDelta := update.QuoteFilled - trade.quoteFilled
+		w.mm.modifyBotCEXBalance(w.botID, trade.toAsset, quoteFilledDelta, balanceModIncrease)
+		trade.quoteFilled = update.QuoteFilled
+		trade.baseFilled = update.BaseFilled
+	}
+
+	if !trade.sell && update.BaseFilled > trade.baseFilled {
+		baseFilledDelta := update.BaseFilled - trade.baseFilled
+		w.mm.modifyBotCEXBalance(w.botID, trade.toAsset, baseFilledDelta, balanceModIncrease)
+		trade.baseFilled = update.BaseFilled
+		trade.quoteFilled = update.QuoteFilled
+	}
+
+	if !update.Complete {
+		return
+	}
+
+	if trade.sell && trade.qty > trade.baseFilled {
+		unfilledQty := trade.qty - trade.baseFilled
+		w.mm.modifyBotCEXBalance(w.botID, trade.fromAsset, unfilledQty, balanceModIncrease)
+	}
+
+	if !trade.sell && calc.BaseToQuote(trade.rate, trade.qty) > trade.quoteFilled {
+		unfilledQty := calc.BaseToQuote(trade.rate, trade.qty) - trade.quoteFilled
```

Comment:
Could this trade have been completed at different rates? If taker maybe some quote was taken at a better rate?

Reply:
The original reserved amount was `calc.BaseToQuote(trade.rate, trade.qty)`. The amount of quote that was used is `trade.quoteFilled`. `quoteFilled` is the amount of quote that was used to buy the base in this trade, regardless of what the rate was.